tls/extensions/s2n_client_key_share.c

```diff
@@ -129,18 +180,27 @@ static int s2n_send_hrr_keyshare(struct s2n_connection *conn, struct s2n_stuffer
     return S2N_SUCCESS;
 }
 
-static int s2n_ecdhe_supported_curves_send(struct s2n_connection *conn, struct s2n_stuffer *out)
-{
-    /* From https://tools.ietf.org/html/rfc8446#section-4.1.2
-     * If a "key_share" extension was supplied in the HelloRetryRequest,
-     * replace the list of shares with a list containing a single
-     * KeyShareEntry from the indicated group.*/
-    if (s2n_is_hello_retry_handshake(conn)) {
-        GUARD(s2n_send_hrr_keyshare(conn, out));
-        return S2N_SUCCESS;
+static int s2n_send_hrr_pq_hybrid_keyshare(struct s2n_connection *conn, struct s2n_stuffer *out) {
+    notnull_check(conn);
+    notnull_check(out);
+
+    S2N_ERROR(S2N_ERR_UNIMPLEMENTED);
+}
+
+/* From https://tools.ietf.org/html/rfc8446#section-4.1.2
+ * If a "key_share" extension was supplied in the HelloRetryRequest,
+ * replace the list of shares with a list containing a single
+ * KeyShareEntry from the indicated group.*/
+static int s2n_send_hrr_keyshare(struct s2n_connection *conn, struct s2n_stuffer *out) {
+    notnull_check(conn);
+    notnull_check(out);
+
+    if (conn->secure.server_kem_group_params.kem_group != NULL) {
+        GUARD(s2n_send_hrr_pq_hybrid_keyshare(conn, out));
```

Comment:
Am I understanding this code correctly that an s2n server with PQ KEM's configured will close the connection instead of attempting to a HRR fall back to ECDHE?

Reply:
No, this doesn't say anything about s2n's server behavior. This `s2n_send_hrr_keyshare()` function executes on the client if the client receives an HRR from the server. If the server chose to negotiate a PQ group, it would have indicated that in the HRR, and when processing the HRR, the client would have set `conn->secure.server_kem_group_params.kem_group` to the KEM group indicated by the server:

https://github.com/awslabs/s2n/blob/89fa7ecc41fcd37ee12a94f7bfe310e08c3213d9/tls/extensions/s2n_server_key_share.c#L85-L92

This is analogous to what happens for if the server chose ECDHE:

https://github.com/awslabs/s2n/blob/89fa7ecc41fcd37ee12a94f7bfe310e08c3213d9/tls/extensions/s2n_server_key_share.c#L157-L164